### PR TITLE
added spot_price value for node_pools

### PR DESCRIPTION
**Pull request by lzecca78**
This is the third missing piece from the other 2 pr I did (https://github.com/sighupio/furyctl/pull/43 and https://github.com/sighupio/fury-kubernetes-aws/pull/51), that allows covering all the spots behind the spot-instance capability feature.
Feedbacks are welcome! 🙏

### CODE REVIEW

File: example/main.tf [hunk 1]
```diff
@@ -55,3 +55,8 @@ users:
         - "${var.cluster_name}"
 EOT
 }
+
+
+output "node_pools" {
```
Comment (angelbarrera92) [inline example/main.tf:60]: Remove this piece of code if is not used

File: modules/eks/output.tf [hunk 2]
```diff
@@ -32,3 +32,8 @@ output "eks_worker_security_group_id" {
   description = "Security group ID attached to the EKS workers."
   value       = module.cluster.worker_security_group_id
 }
+
+
+output "parsed_node_pool" {
```
Comment (angelbarrera92) [inline modules/eks/output.tf:37]: Same here

**Comment by angelbarrera92** [comment]
I suggest to think about the interface of the new "spot" feature.

Would it makes sense something like:

```
node_pools = [
{
min_size = 1
max_size = 2
...
spot = {
  enable = true # Require in EKS, GKE, AKS
  price = 0.11231223 # Require only in EKS
}
}
]
```

I know in azure and gke they don't need the `price` but it required to configure the right attribute. This way, we can maintain the same interface across cloud (aks,gke,eks) installers.

What do you think?

**Comment by nutellinoit** [comment]
Is price mandatory? Usually you use the AWS default, see https://github.com/sighupio/fury-kubernetes-aws/blob/master/modules/aws-kubernetes/kube-worker-spot.tf (there is no price here).

If you set the price, and aws increases the price, your instances will no more start up

**Comment by lzecca78** [comment]
@angelbarrera92 ... I really don't get it... the only entry point in fury for piloting the setup is the yaml, so, why make a "common interface" also in Terraform that is the implementation? It makes no sense because in the implementation you only leave the real needed parameter.... No need to maintain parameters that the implementation doesn't need

**Comment by lzecca78** [comment]
@nutellinoit, we are using in our module the babenko eks official module: https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/spot-instances.md#using-launch-configuration .
Our implementation follows that launch configuration setup, so the only way to do that is to pass the spot_price, as the documentation says

**Comment by angelbarrera92** [comment]
> @angelbarrera92 ... I really don't get it... the only entry point in fury for piloting the setup is the yaml, so, why make a "common interface" also in Terraform that is the implementation? It makes no sense because in the implementation you only leave the real needed parameter.... No need to maintain parameters that the implementation doesn't need

Because this was the main design decision we made with the cloud installers.

**Comment by nutellinoit** [comment]
> @nutellinoit, we are using in our module the babenko eks official module: https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/spot-instances.md#using-launch-configuration .
> Our implementation follows that launch configuration setup, so the only way to do that is to pass the spot_price, as the documentation says

Understood. What happens when the spot instances costs more than the price set?

**Comment by lzecca78** [comment]
>Because this was the main design decision we made with the cloud installers.
Ah ok

**Comment by lzecca78** [comment]
@nutellinoit :
> The spot_price is set to the on-demand price so that the spot instances will run as long as they are the cheaper.

As far as I understood, (and this is what I know regarding the spot instances), you'll be notified with a notification/event and that the instance will be shut down.
This is the nature of the spot instance. But for this reason, I made the deployment of the aws-node-termination-handler: https://github.com/sighupio/fury-kubernetes-aws/pull/51

**Review by lnovara — APPROVED** [review]
@lzecca78 If I understand correctly, to create spot instances for a given node pool I just need to specify the `spot_instance_price` parameter and I can leave it empty to get the "default" price, right?

**Comment by lzecca78** [comment]
mmmm this is not how the babenko terraform module works afaik: https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/spot-instances.md#using-launch-configuration as you can see here @lnovara, the discriminant parameter is to set a value in the parameter `spot_price` otherwise it will be considered as a normal instance. So going back to the furyctl implementation: when you need to inject the spot instance parameter: https://github.com/sighupio/furyctl/pull/43/files#diff-9dfba49c7818625d480d2ebe3e00b2ef393faa4fbb30aa3bd9f55ef53416cb39R39 there is a struct that need to filled up in both its parameters (bool and price) when you're on EKS, while for GKE you only need the boolean.

**Review by nutellinoit — CHANGES_REQUESTED** [review]
Can we put a default price on spot instance, let's say `100` euros per hour, if we enable nodegroup as spot?
I think AWS will still use the lower price possible for the instances.

Or better, can we get the price with a `data` from the instance sizing?

I would like to avoid putting `spot_instance_price`

File: modules/eks/eks.tf [hunk 3]
```diff
@@ -65,6 +70,8 @@ module "cluster" {
       root_volume_size              = lookup(node_pool, "volume_size")
       key_name                      = aws_key_pair.nodes.key_name
       public_ip                     = false
+      // adding 20% more of get price in order to avoid price volatily
```
Comment (nutellinoit) [inline modules/eks/eks.tf:73]: We can remove this comment here

**Review by nutellinoit — APPROVED** [review]
Very little fix, and the changes are now LGTM ❤️ ❤️ ❤️ ❤️ ❤️

**Pull request merged**